### DISCUSSION TITLE
Have a webpage as media?

### DISCUSSION THREAD

**Discussion by Hi-there-how-are-u** [sircharlo/meeting-media-manager#663] [Ideas] [answered]
Would it be possible to show webpages with M³?
For instance, show how the donate.jw.org webpage works, that would be great.
I am not sure what it's called but maybe web app or something when the webpage is shown without the adress and other stuff would be even greater. It would give a clean look. Also during field service with zoom would be great when showing maps and stuff.

Have a great day and thank you for your great work!

**Comment by sircharlo** [comment]
Well well, _someone_ sure sounds like they've used Soundbox in the past. 😆

**Reply by mtdvlpr** [reply]
Soundbox? What might that be?

**Reply by Hi-there-how-are-u** [reply]
I cannot neither confirm or deny that... :) Maybe just maybe I used to do the swedish stuff for soundbox... :)

**Reply by sircharlo** [reply]
One thing I never liked about "other solutions" was how they required moderators, or curators, to manage the media lists. That was one big motivator for me to develop my own bash script, one that didn't depend on people to manually do the grunt work of updating media lists every month (this script is actually the great-great-great grandfather of the current M3 app). I didn't like relying on a third-party to fetch the correct media.

**Reply by sircharlo** [reply]
Other than that though, it really was a nice app.

**Comment by mtdvlpr** [comment]
Hmm, interesting idea! We also have to show jw.org at times which is still quite the hassle...I am definitely interested in digging into this!

**Reply by sircharlo** [reply]
In Soundbox, <s>as we all remember,</s> you would drop a URL shortcut in the media folder, and then you'd be able to view that web page. 
The page itself would show on the TVs, while the same site would show up in a window on the controller side of things as well. 
From the controller, you could navigate the web site, change pages, zoom the webview, scroll the pages, etc., and it would all be mirrored on the TVs.
It was pretty slick.

**Comment by mtdvlpr** [comment]
Hi there, @Hi-there-how-are-u, so it turns out jw.org won't allow you to open it from an external application (which makes sense, since they prevent tempering with the website that way). Since jw.org would be the primary purpose for this feature and it seems like a lot of work to implement, I'm afraid it's a no go.
<details><summary>See image</summary>

![Screenshot from 2022-11-09 15-30-10](https://user-images.githubusercontent.com/46671786/200857923-f63b2ef7-fc1c-4cdf-8a50-9fc7fcd74554.png)


</details>

**Reply by sircharlo** [reply]
Are you sure? I'm pretty sure I've done something similar in the past.

**Reply by mtdvlpr** [reply]
Well, I've tried an iframe, Electron BrowserView and Electron BrowserWindow. All of whom return the error shown in the image above. If you figure out a way to display the website on the media window, I'd be more than happy to continue working on this feature!

**Reply by sircharlo** [reply]
I can't claim to understand why, but it has to do with cookies.

![image](https://user-images.githubusercontent.com/1275666/200900467-25c7397b-2733-471c-ac23-cb02f856e901.png)

**Reply by sircharlo** [reply]
Quick and dirty mod to `index.js` as proof-of-concept:
```
import { app, session, ipcMain, nativeTheme, screen } from 'electron' // import session from electron
....................
      // mediaWinHandler.loadPage('/media')
      const filter = {
        urls: ['*://*.jw.org/*']
      };    
      session.defaultSession.webRequest.onBeforeSendHeaders(
        filter,
        (details, callback) => {
          details.requestHeaders = {
            cookie: "ckLang=E;",
          };
          callback({ requestHeaders: details.requestHeaders });
        }
      );
      mediaWin.loadURL('https://jw.org')
```

**Reply by mtdvlpr** [reply]
Nice!

**Comment by e-Feitosa** [comment]
I always wanted this function 😁, but as I had already asked for many things, I didn't say anything. Would it be something similar to who has in OBS Studio? Here I use like this.

![image](https://user-images.githubusercontent.com/90984283/201087439-a4318ee0-b8fd-4d98-bb2c-5fbbb8dbd06f.png)

**Comment by Chipperchap** [comment]
Hi @e-Feitosa,
Just a query please... while I understand OBS can fetch and display JW.ORG (for example), and this can / will automatically display on ZOOM, how do you navigate the site if it is moved to the auditorium screens / stage screens at the Kingdom Hall so that the in-person friends can see it? 
Or are you able to have some kind of a mirror function where it displays on the big screens while simultaneously being able to navigate around the site on your desktop? Or do you move the browser fetch to the stage screens and navigate around up there while viewing it in the preview of OBS so that you can see what your mouse/cursor is doing?
Our auditorium is very large and moving the JW.ORG site to the stage screens, then seeing the cursor/mouse from a distance is not easy for my aged eyes... despite expanding the size of the cursor/mouse. 

***

My separate request is similar but different (probably should open another topic) - I would like to be able to show our JWLib app on the stage screens while navigating around on the windows 10/11 desktop. This way we can show various functions and features 'live' while the friends try it on their devices / phones etc. (For example the tutorial we had for the meeting some weeks ago on the features of the 'Enjoy' brochure media and endnote etc sections.)  A similar thing happened last week where the brother conducting the assembly preview spontaneously said "Booth, can we please have the assembly program shown on the screens?"  I had to indicate sorry, not able to do that... (or at least not at instant short notice! 😣😁)
 
The only thing that concerns me is that we are not supposed to show custom 'moving pictures' on the stage screens. Does navigating around JW.ORG or potentially JWLib app constitute 'moving images'? (No I am not suggesting a discussion on this, just thinking out 'loud').  To date I have either always taken screenshots and done 'mark-up's' with arrows defining where to navigate to (therefore these are 'static' or still images), or verbally described where to find the sections in JWLib as the friends follow along on their devices... but only because I have not figured out an easy way to display JW.ORG or JWLib on the screens so that the friends can view and follow along.

**Reply by e-Feitosa** [reply]
Hi @Chipperchap, 
As it's not something that happens often and not everyone will be able to run it via OBS Studio, it's best to wait if there will be any M3 features for this. In the case of program, you can download the PDF and import to M3 as you do with the images and then display on the monitor and zoom.

**Comment by mtdvlpr** [comment, marked as answer]
Hi everyone, we're excited to say that we have a working version for this where you can open jw.org directly from the presentation mode! It will be in the next release :)

**Reply by Chipperchap** [reply]
Need any testers...?

**Reply by mtdvlpr** [reply]
Testing is always welcome :)

https://drive.google.com/drive/folders/1_2CzKieYGaK3LcYrEwzNQgfTpNReOSx9

**Reply by Chipperchap** [reply]
Video: [HERE](https://capture.dropbox.com/zquPaTSfQRIgBRBn)

Sorry - didn't realize I had such a noisy mouse!
I don't seem to be able to resize the webpage (which is why I am clicking off screen in the beginning of video, and the cookies alert appears on each page. It doesn't appear on my normal website browsers.

**Reply by mtdvlpr** [reply]
Yeah, that's because jw.org creates a cookie to save your preference, but we can't access that cookie from m3, so each refresh jw.org thinks we're completely new to the page.

**Reply by mtdvlpr** [reply]
And the resizing is because we set quite a large minimum size, to ensure the controller screen is the same as the media screen.

**Reply by Chipperchap** [reply]
> And the resizing is because we set quite a large minimum size, to ensure the controller screen is the same as the media screen.

All good - I was just trying to fit it into a predefined video screen capture area so no problem in a real/live setting. 😊

> Yeah, that's because jw.org creates a cookie to save your preference, but we can't access that cookie from m3, so each refresh jw.org thinks we're completely new to the page.

Oh, I see. So if we remain on the one page after dismissing the cookie request, we will be okay. But if we navigate around the website then we are going to have the banner appear repeatedly... I guess we can be come adept at clicking "accept" quickly to remove the banner from interrupting the view... 🖱 🖱 🖱

**Reply by mtdvlpr** [reply]
I will take another look at it to see if there is some hacky workaround for it, but haven't had any luck so far.

**Reply by mtdvlpr** [reply]
Turns out it was much easier than I thought! The banner will never be shown now :)
